src/service-worker.ts

```diff
@@ -0,0 +1,153 @@
+ * This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+import {writeFile} from 'fs';
+import * as path from 'path';
+import * as logging from 'plylog';
+import {PolymerProject} from './polymer-project';
+import {DepsIndex} from './analyzer';
+const swPrecache = require('sw-precache');
+
+// non-ES compatible modules
+let logger = logging.getLogger('polymer-build.service-worker');
+
+export interface SWConfig {
+  cacheId?: string;
+  directoryIndex?: string;
+  dynamicUrlToDependencies?: {
+    [property: string]: string[]
+  };
+  handleFetch?: boolean;
+  ignoreUrlParametersMatching?: RegExp[];
+  importScripts?: string[];
+  logger?: Function;
+  maximumFileSizeToCacheInBytes?: number;
+  navigateFallback?: string;
+  navigateFallbackWhitelist?: RegExp[];
+  replacePrefix?: string;
+  runtimeCaching?: {
+    urlPattern: RegExp;
+    handler: string;
+    options?: {
+      cache: {
+        maxEntries: number;
+        name: string;
+      };
+    };
+  }[];
+  staticFileGlobs?: string[];
+  stripPrefix?: string;
+  templateFilePath?: string;
+  verbose?: boolean;
+}
+
+export interface AddServiceWorkerOptions {
+  project: PolymerProject;
+  buildRoot: string;
+  bundled?: boolean;
+  serviceWorkerPath?: string;
+  swConfig?: SWConfig;
+}
+
+function getPrecachedAssets(depsIndex: DepsIndex, project: PolymerProject) {
+  let precachedAssets = new Set<string>(project.analyzer.allFragments);
+  precachedAssets.add(project.entrypoint);
+
+  for (let dep of depsIndex.depsToFragments.keys()) {
+    precachedAssets.add(dep);
+  }
+  for (let depImports of depsIndex.fragmentToFullDeps.values()) {
+    depImports.scripts.forEach((s) => precachedAssets.add(s));
+    depImports.styles.forEach((s) => precachedAssets.add(s));
+  }
+  return Array.from(precachedAssets);
+}
+
+function getBundledPrecachedAssets(project: PolymerProject) {
+  let precachedAssets = new Set<string>(project.analyzer.allFragments);
+  precachedAssets.add(project.entrypoint);
+
+  return Array.from(precachedAssets).concat(project.bundler.sharedBundleUrl);
+}
+
+/**
+ * Returns a service worker transform stream. This stream will add a service
+ * worker to the build stream, based on the options passed and build analysis
+ * performed eariler in the stream.
+ *
+ * Note that this stream closely resembles a pass-through stream. It does not
+ * modify the files that pass through it. It only ever adds 1 file.
+ */
+export function generateServiceWorker(options: AddServiceWorkerOptions): Promise<Buffer> {
+  console.assert(!!options.project, '`project` option is required');
```

Comment:
I prefer `options.project != null` for being more explicit and intent preserving, but then again I'm a little weird...